pkg/registry/core/service/storage/rest_test.go

```diff
@@ -3168,242 +2086,141 @@ func TestServiceUpgrade(t *testing.T) {
+		},
+
+		svc: svctest.MakeService("foo", func(s *api.Service) {
+			s.Spec.IPFamilies = []api.IPFamily{api.IPv4Protocol}
+		}),
+	}, {
+		name:                     "upgrade to v4,6",
+		enableDualStackAllocator: true,
+		enableDualStackGate:      true,
+		allocateIPsBeforeUpdate:  nil,
+		expectUpgradeError:       false,
+
+		updateFunc: func(s *api.Service) {
+			s.Spec.IPFamilyPolicy = &requireDualStack
+			s.Spec.IPFamilies = []api.IPFamily{api.IPv4Protocol, api.IPv6Protocol}
+		},
+
+		svc: svctest.MakeService("foo", func(s *api.Service) {
+			s.Spec.IPFamilies = []api.IPFamily{api.IPv4Protocol}
+		}),
+	}, {
+		name:                     "upgrade to v4,6 (specific ip)",
+		enableDualStackAllocator: true,
+		enableDualStackGate:      true,
+		allocateIPsBeforeUpdate:  nil,
+		expectUpgradeError:       false,
+
+		updateFunc: func(s *api.Service) {
+			s.Spec.IPFamilyPolicy = &requireDualStack
+			s.Spec.ClusterIPs = append(s.Spec.ClusterIPs, "2000:0:0:0:0:0:0:1")
+			s.Spec.IPFamilies = []api.IPFamily{api.IPv4Protocol, api.IPv6Protocol}
+		},
+
+		svc: svctest.MakeService("foo", func(s *api.Service) {
+			s.Spec.IPFamilies = []api.IPFamily{api.IPv4Protocol}
+		}),
+	}, {
+		name:                     "upgrade to v4,6 (specific ip) - fail, ip is not available",
+		enableDualStackAllocator: true,
+		enableDualStackGate:      true,
+		allocateIPsBeforeUpdate:  map[api.IPFamily]string{api.IPv6Protocol: "2000:0:0:0:0:0:0:1"},
+		expectUpgradeError:       true,
+
+		updateFunc: func(s *api.Service) {
+			s.Spec.IPFamilyPolicy = &requireDualStack
+			s.Spec.ClusterIPs = append(s.Spec.ClusterIPs, "2000:0:0:0:0:0:0:1")
+			s.Spec.IPFamilies = []api.IPFamily{api.IPv4Protocol, api.IPv6Protocol}
+		},
+
+		svc: svctest.MakeService("foo", func(s *api.Service) {
+			s.Spec.IPFamilies = []api.IPFamily{api.IPv4Protocol}
+		}),
+	}, {
+		name:                     "upgrade to v6,4",
+		enableDualStackAllocator: true,
+		enableDualStackGate:      true,
+		allocateIPsBeforeUpdate:  nil,
+		expectUpgradeError:       false,
+
+		updateFunc: func(s *api.Service) {
+			s.Spec.IPFamilyPolicy = &requireDualStack
+			s.Spec.IPFamilies = []api.IPFamily{api.IPv6Protocol, api.IPv4Protocol}
+		},
+
+		svc: svctest.MakeService("foo", func(s *api.Service) {
+			s.Spec.IPFamilies = []api.IPFamily{api.IPv6Protocol}
+		}),
+	}, {
+		name:                     "upgrade to v6,4 (specific ip)",
+		enableDualStackAllocator: true,
+		enableDualStackGate:      true,
+		allocateIPsBeforeUpdate:  nil,
+		expectUpgradeError:       false,
+
+		updateFunc: func(s *api.Service) {
+			s.Spec.IPFamilyPolicy = &requireDualStack
+			s.Spec.ClusterIPs = append(s.Spec.ClusterIPs, "1.2.3.4")
+			s.Spec.IPFamilies = []api.IPFamily{api.IPv6Protocol, api.IPv4Protocol}
+		},
+
+		svc: svctest.MakeService("foo", func(s *api.Service) {
+			s.Spec.IPFamilies = []api.IPFamily{api.IPv6Protocol}
+		}),
+	}, {
+		name:                     "upgrade to v6,4 (specific ip) - fail ip is already allocated",
```

Comment:
I think we can safety remove that. We covered alloc an allocated IP so many times in other tests including dual stack

Reply:
I'm going to defer this sort of cleanup until after delayering